### ISSUE TITLE
Shared dependencies of dependencies resolution question

### ISSUE THREAD

**Issue by otbe** [yarnpkg/yarn#1272]
**Do you want to request a *feature* or report a *bug*?**
bug?

**Please mention your node.js, yarn and operating system version.**
yarn 0.16, node 6.5.0 and OS X 10.12.1

**What is the current behavior?**
```npm``` and ```yarn``` differ in the way how deps of deps are resolved. 

For example:
There is a public module called ```something``` and its latest version is 0.14.41.
module As package.json:
```
{
   "name": "moduleA",
   "version": "1.0.0",
   "dependencies": {
      "something": "^0.14.30"
   }
}
```

module Bs package.json:
```
{
   "name": "moduleB",
   "dependencies": {
      "something": "0.14.40" 
      "moduleA": "^1.0.0"
   }
}
```
```npm``` will install one version (0.14.40) of ```something``` in module Bs node_modules folder, but ```yarn``` will install two versions. 0.14.0 in module Bs and 0.14.41 in module As node_modules folder. 

Compared to npm this leads to unexpected behavior for bundlers and other processing tools (like TS). As a side note: ```something``` is a public scoped package, maybe this is related to the scoping issues.

**What is the expected behavior?**
I like the way how ```npm``` solves this problem :)
I know the ```--flat``` option for ```yarn install```, but I don't want to choose between version ```1.0.0``` and ```2.0.0```of a package. I think ```yarn``` should simply install semver compatible versions as flat as possible. In my example above 0.14.40 matches requirements from both modules (I think this is what ```npm``` does).

Maybe my understanding is wrong, but whats the expected behavior in this case? 
Thanks!

**Comment by davidmerrique** [comment]
I think this is the same as or related to #1158

This is definitely a problem.

**Comment by otbe** [comment]
Indeed! Sorry for duplicate, searched a while for an related issue 👍

**Comment by kittens** [comment]
Fixed via #1477.
